pkg/vm/builtins/ecdsa.go

```diff
@@ -0,0 +1,160 @@
+	ecdsaIndex := offset % cellsPerECDSA
+	pubOffset := offset - ecdsaIndex
+	msg_offset := pubOffset + 1
+
+	pub := segment.Peek(pubOffset)
+	if !pub.Known() {
+		//Not sure if this is the right approach. It seems the msg and pub key  can be passed in either order.
+		return nil
+		//return fmt.Errorf("cannot infer value: input value at offset %d is unknown", pubOffset)
+	}
+
+	pubX, err := pub.FieldElement() //X element of the sig
+	if err != nil {
+		return err
+	}
+
+	msg := segment.Peek(msg_offset)
+	if !msg.Known() {
+		return nil
+		//return fmt.Errorf("cannot infer value: input value at offset %d is unknown", msg_offset)
+	}
+	msgField, err := msg.FieldElement()
+	if err != nil {
+		return err
+	}
+
+	//Recover Y part of the public key
+	posY, negY, err := recoverY(pubX)
+	if err != nil {
+		return err
+	}
+
+	//Try first with positive y
+	key := starkcurve.G1Affine{X: *pubX, Y: *posY}
+	if !key.IsOnCurve() {
+		return fmt.Errorf("Key is not on curve")
+	}
+
+	pubKey := &ecdsa.PublicKey{A: key}
+	sig, ok := e.signatures[pubOffset]
+	if !ok {
+		return fmt.Errorf("Signature is missing form ECDA builtin")
+	}
+
+	msgBytes := msgField.Bytes()
+	valid, err := pubKey.Verify(sig.Bytes(), msgBytes[:], nil)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		// Now try with Neg Y. Already know the point is on the curve so no need to check again
+		key = starkcurve.G1Affine{X: *pubX, Y: *negY}
+		pubKey = &ecdsa.PublicKey{A: key}
+		valid, err := pubKey.Verify(sig.Bytes(), msgBytes[:], nil)
+		if err != nil {
+			return err
+		}
+		if !valid {
+			return fmt.Errorf("Signature is not valid")
+		}
+	}
+	fmt.Println("VALID")
+	return nil
+}
+
+func (e *ECDSA) InferValue(segment *memory.Segment, offset uint64) error {
+	return fmt.Errorf("Can't infer value")
+}
+
+/*
+Hint that will call this function looks like this:
+
+	"hints": {
+	    "6": [
+	        {
+	            "accessible_scopes": [
+	                "starkware.cairo.common.signature",
+	                "starkware.cairo.common.signature.verify_ecdsa_signature"
+	            ],
+	            "code": "ecdsa_builtin.add_signature(ids.ecdsa_ptr.address_, (ids.signature_r, ids.signature_s))",
+	            "flow_tracking_data": {
+	                "ap_tracking": {
+	                    "group": 2,
+	                    "offset": 0
+	                },
+	                "reference_ids": {
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.ecdsa_ptr": 4,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.message": 0,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.public_key": 1,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.signature_r": 2,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.signature_s": 3
+	                }
+	            }
+	        }
+	    ]
+	},
+*/
+func (e *ECDSA) AddSignature(pubOffset uint64, r, s fp.Element) error {
```

Comment:
`r` and `s` should be pointers(`*fp.Element`). The general style is that we receive references and return value types. This way we avoid copying the full Felt, and just a pointer to it. This pointer references the stack and not the heap if this technique is done well (go compiler perform escape analysis)